File: algorithms/0235-lowest-common-ancestor-of-a-bst/Solution.java
// Lowest Common Ancestor of a BST - iterative walk using BST order.
// Time O(h), Space O(1).

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode cur = root;
        while (cur != null) {
            if (p.val > cur.val && q.val > cur.val) {
                cur = cur.right;
            } else if (p.val < cur.val && q.val < cur.val) {
                cur = cur.left;
            } else {
                return cur;
            }
        }
        return null;
    }
}

File: algorithms/0235-lowest-common-ancestor-of-a-bst/solution.py
# Lowest Common Ancestor of a BST - iterative walk using BST order.
# Time O(h), Space O(1).


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode | None, p: TreeNode | None, q: TreeNode | None
    ) -> TreeNode | None:
        cur = root
        while cur is not None:
            if p.val > cur.val and q.val > cur.val:
                cur = cur.right
            elif p.val < cur.val and q.val < cur.val:
                cur = cur.left
            else:
                return cur
        return None
